internal/log/log.go

```diff
@@ -0,0 +1,151 @@
+// Package log provides logging utilities for the tracer.
+package log
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
+)
+
+// Level specifies the logging level that the log package prints at.
+type Level int
+
+const (
+	// LevelDebug represents debug level messages.
+	LevelDebug Level = iota
+	// LevelWarn represents warning and errors.
+	LevelWarn
+)
+
+var prefixMsg = fmt.Sprintf("Datadog Tracer %s", version.Tag)
+
+var (
+	mu     sync.RWMutex   // guards below fields
+	level                 = LevelWarn
+	logger ddtrace.Logger = &defaultLogger{l: log.New(os.Stderr, "", log.LstdFlags)}
+)
+
+// UseLogger sets l as the active logger.
+func UseLogger(l ddtrace.Logger) {
+	mu.Lock()
+	defer mu.Unlock()
+	logger = l
+}
+
+// SetLevel sets the given lvl for logging.
+func SetLevel(lvl Level) {
+	mu.Lock()
+	defer mu.Unlock()
+	level = lvl
+}
+
+// Debug prints the given message if the level is LevelDebug.
+func Debug(fmt string, a ...interface{}) {
+	mu.RLock()
+	lvl := level
+	mu.RUnlock()
+	if lvl != LevelDebug {
+		return
+	}
+	printMsg("DEBUG", fmt, a...)
+}
+
+// Warn prints a warning message.
+func Warn(fmt string, a ...interface{}) {
+	printMsg("WARN", fmt, a...)
+}
+
+var (
+	errmu   sync.RWMutex                // guards below fields
+	erragg  = map[string]*errorReport{} // aggregated errors
+	errrate time.Duration               // the rate at which errors are reported
+	erron   bool                        // true if errors are being aggregated
+)
+
+func init() {
+	errrate = time.Minute
+	if v, ok := os.LookupEnv("DD_LOGGING_RATE"); ok {
+		if sec, err := strconv.Atoi(v); err != nil {
+			Warn("Invalid value for DD_LOGGING_RATE: %v", err)
+		} else {
+			errrate = time.Duration(sec) * time.Second
+		}
+	}
+}
+
+type errorReport struct {
+	err   error
+	count uint64
+}
+
+// Error aggregates errors under the given key. The aggregated errors are printed
+// once a minute or once every DD_LOGGING_RATE number of seconds.
+func Error(key, format string, a ...interface{}) {
+	if reachedLimit(key) {
+		// avoid too much lock contention on spammy errors
+		return
+	}
+	errmu.Lock()
+	defer errmu.Unlock()
+	report, ok := erragg[key]
+	if !ok {
+		erragg[key] = &errorReport{err: fmt.Errorf(format, a...)}
+		report = erragg[key]
+	}
+	report.count++
+	if !erron {
+		erron = true
+		time.AfterFunc(errrate, Flush)
```

Comment:
So errors get reported up to 60s after they occur (depending on when this timer started, and the timing of the error occuring), but warnings get reported immediately?
This seems confusing.

Reply:
That's because there is no actually risk of warnings becoming spammy. We only have a few instances of them. Do you think there's any reasons why we should do the same for errors? It would complicate things a bit. But, if the previous suggestion's experiment succeeds (using caller line/number), then I guess we might as well do it for both.

Reply:
Gonna mark this as resolved for now.